xray/test/test_plotting.py

```diff
@@ -0,0 +1,190 @@
+import numpy as np
+
+from xray import DataArray
+
+from . import TestCase, requires_matplotlib
+
+try:
+    import matplotlib as mpl
+    # Using a different backend makes Travis CI work.
+    mpl.use('Agg')
+    # Order of imports is important here.
+    import matplotlib.pyplot as plt
+except ImportError:
+    pass
+
+
+# TODO - Add NaN handling and tests
+
+@requires_matplotlib
+class PlotTestCase(TestCase):
+
+    def tearDown(self):
+        # Remove all matplotlib figures
+        plt.close('all')
+
+    def pass_in_axis(self, plotfunc):
+        fig, axes = plt.subplots(ncols=2)
+        plotfunc(ax=axes[0])
+        self.assertTrue(axes[0].has_data())
+
+    def imshow_called(self, plotfunc):
+        ax = plotfunc()
+        images = ax.findobj(mpl.image.AxesImage)
+        return len(images) > 0
+
+    def contourf_called(self, plotfunc):
+        ax = plotfunc()
+        paths = ax.findobj(mpl.collections.PathCollection)
+        return len(paths) > 0
+
+
+class TestPlot(PlotTestCase):
+
+    def setUp(self):
+        d = np.arange(24).reshape(2, 3, 4)
+        self.darray = DataArray(d)
+
+    def test1d(self):
+        self.darray[0, 0, :].plot()
+
+    def test2d_uniform_calls_imshow(self):
+        a = self.darray[0, :, :]
+        self.assertTrue(self.imshow_called(a.plot))
+
+    def test2d_nonuniform_calls_contourf(self):
+        a = self.darray[0, :, :]
+        a.coords['dim_1'] = [0, 10, 2]
+        self.assertTrue(self.contourf_called(a.plot))
+
+    def test3d(self):
+        self.darray.plot()
+
+    def test_can_pass_in_axis(self):
+        self.pass_in_axis(self.darray.plot)
+
+
+class TestPlot1D(PlotTestCase):
+
+    def setUp(self):
+        d = [0, 1, 0, 2]
+        self.darray = DataArray(d, coords={'period': range(len(d))})
+
+    def test_xlabel_is_index_name(self):
+        ax = self.darray.plot()
+        self.assertEqual('period', ax.get_xlabel())
+
+    def test_no_label_name_on_y_axis(self):
+        ax = self.darray.plot()
+        self.assertEqual('', ax.get_ylabel())
+
+    def test_ylabel_is_data_name(self):
+        self.darray.name = 'temperature'
+        ax = self.darray.plot()
+        self.assertEqual(self.darray.name, ax.get_ylabel())
+
+    def test_wrong_dims_raises_valueerror(self):
+        twodims = DataArray(np.arange(10).reshape(2, 5))
+        with self.assertRaises(ValueError):
+            twodims.plot_line()
+
+    def test_format_string(self):
+        self.darray.plot_line('ro')
+
+    def test_can_pass_in_axis(self):
+        self.pass_in_axis(self.darray.plot_line)
+
+    def test_nonnumeric_index_raises_typeerror(self):
+        a = DataArray([1, 2, 3], {'letter': ['a', 'b', 'c']})
+        with self.assertRaisesRegexp(TypeError, r'[Ii]ndex'):
+            a.plot_line()
+
+
```

Comment:
Might be worth defining a 2d mixin with common 2d tests and then make classes for each 2d function, ie: `TestImshow`, ...